### PR TITLE
Adding HIS Token generation in Rust

### PR DESCRIPTION
**Pull request by suvamM**
This PR makes the following changes:

1. Refactors the `microsoft_security_utilties` Rust library in order to cleanly seperate the tests from the Marvin checksum and token generation code.

2. Ports the [C# HIS token generation](https://github.com/microsoft/security-utilities/blob/ef3374b51343f036a897e222061312d6304c2ea2/src/Microsoft.Security.Utilities/IdentifiableSecrets.cs#L53) algorithm into Rust.

3. Ports some the C# token generation tests to Rust. Since the C# uses some .Net specific libraries in the test code, not all tests have been ported. To work around this, I generated secrets in Rust, and verified in .Net. Porting the comprehensive set of tests over is critical future work.

### CODE REVIEW

File: src/Microsoft.Security.Utilities/IdentifiableSecrets.cs [hunk 1]
```diff
@@ -239,58 +239,58 @@ public static bool ValidateBase64Key(string key, ulong checksumSeed, string base
...
-                                 firstChar == 'C' || firstChar == 'D' ||
-                                 firstChar == 'E' || firstChar == 'F' ||
-                                 firstChar == 'G' || firstChar == 'H' ||
-                                 firstChar == 'I' || firstChar == 'J' ||
-                                 firstChar == 'K' || firstChar == 'L' ||
-                                 firstChar == 'M' || firstChar == 'N' ||
-                                 firstChar == 'O' || firstChar == 'P');
-                    break;
-                }
+                    {
```
Comment (suvamM) [inline src/Microsoft.Security.Utilities/IdentifiableSecrets.cs:242]: These changes seem to be due to some line ending issues.

File: src/microsoft_security_utilities/Cargo.toml [hunk 2]
```diff
@@ -5,4 +5,11 @@ edition = "2021"
 
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
-[dependencies]
\ No newline at end of file
+[dependencies]
```
Comment (suvamM) [inline src/microsoft_security_utilities/Cargo.toml:8]: Dependencies for the Rust library.

File: src/microsoft_security_utilities/src/identifiable_secrets_tests.rs [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+
+#[cfg(test)]
+use super::*;
+use std::{collections::HashSet, hash::Hash};
+use base64::alphabet;
+use chrono::prelude::*;
+use rand::prelude::*;
+use rand_chacha::ChaCha20Rng;
+
+static S_BASE62_ALPHABET: &str = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890";
```
Comment (suvamM) [inline src/microsoft_security_utilities/src/identifiable_secrets_tests.rs:12]: The following tests are ported from [here](https://github.com/microsoft/security-utilities/blob/main/src/Tests.Microsoft.Security.Utilities/IdentifiableSecretsTests.cs).

File: src/microsoft_security_utilities/src/identifiable_secrets_tests.rs [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+    }
+    else
+    {
+        set.insert('+');
+        set.insert('/');
+    }
+    set
+}
+
+fn validate_secret(secret: String, seed: u64, signature: String, encode_for_url: bool)
```
Comment (suvamM) [inline src/microsoft_security_utilities/src/identifiable_secrets_tests.rs:277]: This function is a much shorter version of the corresponding [C# test](https://github.com/microsoft/security-utilities/blob/ef3374b51343f036a897e222061312d6304c2ea2/src/Tests.Microsoft.Security.Utilities/IdentifiableSecretsTests.cs#L215). The reason is the function uses .Net specific libraries which either need a Rust port, or a search for Rust equivalent. As a workaround, I emitted the secrets generated in Rust in a `csv`, and then consumed and verified them using the C# function.

File: src/microsoft_security_utilities/src/lib.rs [hunk 5]
```diff
@@ -3,170 +3,5 @@
...
-    #[test]
-    #[should_panic]
-    fn marvin_compute_hash_panic_if_invalid_args_4()
-    {
-        let input = "".as_bytes();
-        microsoft_security_utilities::marvin::compute_hash(input, 0, 3, 3);
-    }
-
-}
+mod marvin_tests;
```
Comment (suvamM) [inline src/microsoft_security_utilities/src/lib.rs:6]: Refactoring the library to keep the main library file clean.

File: src/microsoft_security_utilities/src/microsoft_security_utilities/identifiable_secrets.rs [hunk 6]
```diff
@@ -0,0 +1,496 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+use base64::{engine::general_purpose, Engine as _};
```
Comment (suvamM) [inline src/microsoft_security_utilities/src/microsoft_security_utilities/identifiable_secrets.rs:4]: Rust port of [C# implementation](https://github.com/microsoft/security-utilities/blob/main/src/Microsoft.Security.Utilities/IdentifiableSecrets.cs).

File: src/microsoft_security_utilities/src/identifiable_secrets_tests.rs (hunk 3 above)
Comment (michaelcfanning) [inline src/microsoft_security_utilities/src/identifiable_secrets_tests.rs:12]: >static [](http://example.com/codeflow?start=0&length=6)

too many blank lines

File: src/microsoft_security_utilities/src/identifiable_secrets_tests.rs [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+    }
+}
+
+#[test]
+fn identifiable_secrets_generate_standard_base64_key_should_throw_exception_for_invalid_lengths()
+{        
+    let signature = "ABCD";
+    let seed = 0;
+    
+    let mut result = std::panic::catch_unwind(||
```
Comment (michaelcfanning) [inline src/microsoft_security_utilities/src/identifiable_secrets_tests.rs:78]: >catch_unwind [](http://example.com/codeflow?start=33&length=12)

Spacing issues.

File: src/microsoft_security_utilities/src/identifiable_secrets_tests.rs [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+        );
+    assert!(result.is_err());
+
+    result = std::panic::catch_unwind(|| 
+        microsoft_security_utilities::identifiable_secrets::generate_standard_safe_base64_key
+        (seed,
+            32,
+            "this signature is too long")
+        );
+    assert!(result.is_err());
```
Comment (michaelcfanning) [inline src/microsoft_security_utilities/src/identifiable_secrets_tests.rs:108]: >assert [](http://example.com/codeflow?start=4&length=6)

scrub for blank lines?

File: src/microsoft_security_utilities/src/identifiable_secrets_tests.rs [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+    let alphabet: HashSet<char> = add_to_set(true);
+    
+    for i in 0..42
+    {
+        let injected_char = unsafe {char::from_u32_unchecked(i)};
+        println!("Testing character {} for {}", injected_char, i);
+        let mut signature: String = String::from("XXX");
+        signature.push(injected_char);
+
+        // If the injected character is legal, we'll go ahead and validate everything works as expected.
```
Comment (michaelcfanning) [inline src/microsoft_security_utilities/src/identifiable_secrets_tests.rs:209]: >If [](http://example.com/codeflow?start=11&length=2)

comment is wrong

File: src/microsoft_security_utilities/src/microsoft_security_utilities/identifiable_secrets.rs [hunk 10]
```diff
@@ -0,0 +1,496 @@
...
+pub static MAXIMUM_GENERATED_KEY_SIZE: u32 = 4096;
+pub static MINIMUM_GENERATED_KEY_SIZE: u32 = 24;
+static BITS_IN_BYTES: i32 = 8;
+static BITS_IN_BASE64_CHARACTER: i32 = 6;
+static SIZE_OF_CHECKSUM_IN_BYTES: i32 = mem::size_of::<u32>() as i32;
+
+pub fn is_base62_encoding_char(ch: char) -> bool
+{
+    return (ch >= 'a' && ch <= 'z') ||
+            (ch >= 'A' && ch <= 'Z') ||
```
Comment (michaelcfanning) [inline src/microsoft_security_utilities/src/microsoft_security_utilities/identifiable_secrets.rs:22]: >( [](http://example.com/codeflow?start=12&length=1)

more spacing nits

**Review by michaelcfanning — APPROVED** [review]
:shipit:

**Comment by michaelcfanning** [comment]
Nice change! What a great contribution.

**Pull request merged**